src/Sulu/Bundle/MediaBundle/Trash/MediaTrashItemHandler.php

```diff
@@ -0,0 +1,390 @@
+                    $formatOption->setFormatKey($formatOptionData['formatKey']);
+                    $fileVersion->addFormatOptions($formatOption);
+
+                    $formatOption->setCropHeight($formatOptionData['cropHeight']);
+                    $formatOption->setCropWidth($formatOptionData['cropWidth']);
+                    $formatOption->setCropX($formatOptionData['cropX']);
+                    $formatOption->setCropY($formatOptionData['cropY']);
+                }
+
+                foreach ($fileVersionData['tagIds'] as $tagId) {
+                    if ($tag = $this->findEntity(TagInterface::class, $tagId)) {
+                        $fileVersion->addTag($tag);
+                    }
+                }
+
+                foreach ($fileVersionData['categoryIds'] as $categoryId) {
+                    if ($category = $this->findEntity(CategoryInterface::class, $categoryId)) {
+                        $fileVersion->addCategory($category);
+                    }
+                }
+
+                foreach ($fileVersionData['targetGroupIds'] as $targetGroupId) {
+                    if ($targetGroup = $this->findEntity(TargetGroupInterface::class, $targetGroupId)) {
+                        $fileVersion->addTargetGroup($targetGroup);
+                    }
+                }
+            }
+        }
+
+        $this->domainEventCollector->collect(
+            new MediaRestoredEvent($media, $data)
+        );
+
+        if (null === $this->mediaRepository->findMediaById($id)) {
+            $this->doctrineRestoreHelper->persistAndFlushWithId($media, $id);
+        } else {
+            $this->entityManager->persist($media);
+            $this->entityManager->flush();
+        }
+
+        return $media;
+    }
+
+    public function remove(TrashItemInterface $trashItem): void
+    {
+        $data = $trashItem->getRestoreData();
+
+        foreach ($data['files'] as $fileData) {
+            foreach ($fileData['fileVersions'] as $fileVersionData) {
+                $trashStorageOptions = $this->getTrashStorageOptions($fileVersionData['storageOptions']);
+                $this->storage->remove($trashStorageOptions);
+            }
+        }
+    }
+
+    public static function getResourceKey(): string
+    {
+        return MediaInterface::RESOURCE_KEY;
+    }
+
+    /**
+     * @template T of object
+     *
+     * @param class-string<T> $className
+     * @param mixed|null $id
+     *
+     * @return T|null
+     */
+    private function findEntity(string $className, $id)
+    {
+        if ($id) {
+            return $this->entityManager->find($className, $id);
+        }
+
+        return null;
+    }
+
+    /**
+     * @param array<string, string|null> $originalStorageOptions
+     *
+     * @return array<string, string|null>
+     */
+    private function getTrashStorageOptions(array $originalStorageOptions): array
+    {
+        return \array_merge($originalStorageOptions, ['directory' => 'trash']);
```

Comment:
we would need to prefix here the `fileName` with a uuid or the fileVersion id to avoid conflicts.

Reply:
we decided to resolve conflicts in the `StorageInterface::move` method instead 🙂